appservice/src/wizard.ts

```diff
@@ -0,0 +1,197 @@
+            step: this.steps[this.steps.length - 1],
+            error: null
+        };
+
+        return this._result;
+    }
+
+    get steps(): WizardStep[] {
+        return this._steps;
+    }
+
+    findStepOfType<T extends WizardStep>(stepTypeConstructor: { new(...args: any[]): T }, isOptional?: boolean): T {
+        return <T>this.findStep(
+            step => step instanceof stepTypeConstructor,
+            isOptional ? null : `The Wizard should have had a ${stepTypeConstructor.name} step`);
+    }
+
+    findStep(predicate: (step: WizardStep) => boolean, errorMessage?: string): WizardStep {
+        const step = this.steps.find(predicate);
+
+        if (!step && errorMessage) {
+            throw new Error(errorMessage);
+        }
+
+        return step;
+    }
+
+    write(text: string) {
+        this.output.append(text);
+    }
+
+    writeline(text: string) {
+        this.output.appendLine(text);
+    }
+
+    private onError(err: Error, step: WizardStep) {
+        if (err instanceof UserCancelledError) {
+            throw err;
+        }
+
+        this.writeline(`Error: ${err.message}`);
+        this.writeline('');
+        throw new WizardFailedError(err, step.telemetryStepTitle, step.stepIndex);
+    }
+
+    protected abstract beforeExecute(step?: WizardStep, stepIndex?: number);
+}
+
+export interface WizardResult {
+    status: WizardStatus;
+    step: WizardStep;
+    error: Error | null;
+}
+
+export interface WizardStatePersistence {
+
+}
+
+export class WizardStep {
+    protected constructor(readonly wizard: WizardBase, readonly telemetryStepTitle: string, private persistenceState?: vscode.Memento) { }
+
+    async prompt(): Promise<void> { }
+    async execute(): Promise<void> { }
+
+    get stepIndex(): number {
+        return this.wizard.steps.findIndex(step => step === this);
+    }
+
+    get stepProgressText(): string {
+        return `Step ${this.stepIndex + 1}/${this.wizard.steps.length}`;
+    }
+
+    async showQuickPick<T>(items: QuickPickItemWithData<T>[] | Thenable<QuickPickItemWithData<T>[]>, options: vscode.QuickPickOptions, persistenceKey?: string, token?: vscode.CancellationToken): Promise<QuickPickItemWithData<T>> {
+        options.ignoreFocusOut = true;
+        var resolvedItems = await items;
+        if (this.persistenceState && persistenceKey) {
+            // See if the previous value selected by the user is in this list, and move it to the top as default
+            var previousId = this.persistenceState.get(persistenceKey);
+            var previousItem = previousId && resolvedItems.find(item => item.persistenceId === previousId);
+            if (previousItem) {
+                resolvedItems = ([previousItem]).concat(resolvedItems.filter(item => item !== previousItem));
+            }
+        }
+
+        const result = await vscode.window.showQuickPick(resolvedItems, options, token);
+        if (!result) {
+            throw new UserCancelledError();
```

Comment:
I like using UserCancelledError inside our own code, but I think we should make sure to catch it and return `undefined` at 'api' points. It's not really an _error_ and shouldn't be exposed like one